File: desktop/src/desktop/shell/catalog.rs
use super::super::model::{
    CameraMatchPreference, Collection, ImportResolution, LibraryManifest, PendingDelete, Photo,
    Preset, Stack, StorageReferences,
};
use super::super::{DesktopState, command_error, with_library, with_library_mut};
use std::collections::HashMap;
use tauri::State;

type CommandResult<T> = std::result::Result<T, String>;

#[tauri::command]
pub fn catalog_load_library(
    state: State<'_, DesktopState>,
) -> CommandResult<Option<LibraryManifest>> {
    with_library(&state, |library| library.load_library()).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_save_library(
    library: LibraryManifest,
    state: State<'_, DesktopState>,
) -> CommandResult<()> {
    with_library_mut(&state, |active| active.save_library(&library)).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_resolve_imports(
    photos: Vec<Photo>,
    state: State<'_, DesktopState>,
) -> CommandResult<ImportResolution> {
    with_library(&state, |library| library.resolve_imports(&photos)).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_add_photos(
    library_created_at: u64,
    photos: Vec<Photo>,
    collection: Option<Collection>,
    state: State<'_, DesktopState>,
) -> CommandResult<()> {
    with_library_mut(&state, |library| {
        library.add_photos(library_created_at, &photos, collection.as_ref())
    })
    .map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_update_photo(photo: Photo, state: State<'_, DesktopState>) -> CommandResult<()> {
    with_library_mut(&state, |library| library.update_photo(&photo)).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_save_collection(
    collection: Collection,
    state: State<'_, DesktopState>,
) -> CommandResult<()> {
    with_library_mut(&state, |library| library.save_collection(&collection)).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_delete_collection(
    collection_id: String,
    state: State<'_, DesktopState>,
) -> CommandResult<()> {
    with_library_mut(&state, |library| library.delete_collection(&collection_id))
        .map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_save_stacks(
    stacks: Vec<Stack>,
    changed_photos: HashMap<String, Option<String>>,
    state: State<'_, DesktopState>,
) -> CommandResult<()> {
    with_library_mut(&state, |library| {
        library.save_stacks(&stacks, &changed_photos)
    })
    .map_err(command_error)
}

#[tauri::command]
pub fn catalog_list_presets(state: State<'_, DesktopState>) -> CommandResult<Vec<Preset>> {
    with_library(&state, |library| library.list_presets()).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_save_preset(preset: Preset, state: State<'_, DesktopState>) -> CommandResult<()> {
    with_library(&state, |library| library.save_preset(&preset)).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_delete_preset(
    preset_id: String,
    state: State<'_, DesktopState>,
) -> CommandResult<()> {
    with_library(&state, |library| library.delete_preset(&preset_id)).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_save_camera_match_preference(
    preference: CameraMatchPreference,
    state: State<'_, DesktopState>,
) -> CommandResult<()> {
    with_library(&state, |library| {
        library.save_camera_match_preference(preference)
    })
    .map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_delete_photo(
    photo_id: String,
    render_cache_name: String,
    state: State<'_, DesktopState>,
) -> CommandResult<Vec<PendingDelete>> {
    with_library_mut(&state, |library| {
        library.delete_photo(&photo_id, render_cache_name)
    })
    .map_err(command_error)
}

#[tauri::command]
pub fn catalog_pending_deletions(
    state: State<'_, DesktopState>,
) -> CommandResult<Vec<PendingDelete>> {
    with_library(&state, |library| library.pending_deletions()).map_err(command_error)
}

#[tauri::command(rename_all = "camelCase")]
pub fn catalog_complete_deletions(
    deletions: Vec<PendingDelete>,
    state: State<'_, DesktopState>,
) -> CommandResult<()> {
    with_library_mut(&state, |library| library.complete_deletions(&deletions))
        .map_err(command_error)
}

#[tauri::command]
pub fn catalog_storage_references(
    state: State<'_, DesktopState>,
) -> CommandResult<StorageReferences> {
    with_library(&state, |library| library.storage_references()).map_err(command_error)
}

File: web/src/lib/desktop/catalog.ts
import { invoke } from '@tauri-apps/api/core';
import type { LibraryStack } from '../library-catalog.ts';
import {
	libraryManifestSchema,
	photoCollectionSchema,
	storedPhotoSchema,
	type LibraryManifest,
	type PhotoCollection,
	type StoredPhoto
} from '../library-schema.ts';
import { presetSchema, type Preset } from '../preset.ts';

export interface PendingDelete {
	kind: 'original' | 'thumbnail' | 'edit' | 'derived';
	storageName: string;
	queuedAt: number;
}

export interface StorageReferences {
	originals: string[];
	thumbnails: string[];
	edits: string[];
	masks: string[];
	photoIds: string[];
}

interface ImportResolution {
	additions: StoredPhoto[];
	photoIds: Record<string, string>;
}

export function createDesktopCatalog() {
	return {
		async loadLibrary() {
			const library = await invoke<LibraryManifest | null>('catalog_load_library');
			return library ? libraryManifestSchema.parse(library) : null;
		},

		async resolveImports(photos: readonly StoredPhoto[]) {
			const resolution = await invoke<ImportResolution>('catalog_resolve_imports', { photos });
			return {
				additions: resolution.additions.map((photo) => storedPhotoSchema.parse(photo)),
				photoIds: resolution.photoIds
			};
		},

		saveLibrary(library: LibraryManifest) {
			return invoke<void>('catalog_save_library', { library: nativeManifest(library) });
		},

		addPhotos(
			libraryCreatedAt: number,
			photos: readonly StoredPhoto[],
			collection: PhotoCollection | null
		) {
			return invoke<void>('catalog_add_photos', {
				libraryCreatedAt,
				photos,
				collection: collection ? nativeCollection(collection) : null
			});
		},

		updatePhotoState(photo: StoredPhoto) {
			return invoke<void>('catalog_update_photo', { photo: storedPhotoSchema.parse(photo) });
		},

		saveCollection(collection: PhotoCollection) {
			return invoke<void>('catalog_save_collection', {
				collection: nativeCollection(photoCollectionSchema.parse(collection))
			});
		},

		deleteCollection(collectionId: string) {
			return invoke<void>('catalog_delete_collection', { collectionId });
		},

		saveStacks(stacks: readonly LibraryStack[], changedPhotos: ReadonlyMap<string, string | null>) {
			return invoke<void>('catalog_save_stacks', {
				stacks,
				changedPhotos: Object.fromEntries(changedPhotos)
			});
		},

		async listPresets() {
			return (await invoke<Preset[]>('catalog_list_presets')).map((preset) =>
				presetSchema.parse(preset)
			);
		},

		savePreset(preset: Preset) {
			return invoke<void>('catalog_save_preset', { preset: presetSchema.parse(preset) });
		},

		deletePreset(presetId: string) {
			return invoke<void>('catalog_delete_preset', { presetId });
		},

		deletePhoto(photoId: string, renderCacheName: string) {
			return invoke<PendingDelete[]>('catalog_delete_photo', { photoId, renderCacheName });
		},

		pendingDeletions() {
			return invoke<PendingDelete[]>('catalog_pending_deletions');
		},

		completeDeletions(deletions: readonly PendingDelete[]) {
			return invoke<void>('catalog_complete_deletions', { deletions });
		},

		storageReferences() {
			return invoke<StorageReferences>('catalog_storage_references');
		},

		close() {
			void invoke('close_library');
		}
	};
}

export type DesktopCatalog = ReturnType<typeof createDesktopCatalog>;

function nativeManifest(library: LibraryManifest) {
	return {
		...library,
		collections: library.collections.map(nativeCollection)
	};
}

function nativeCollection(collection: PhotoCollection) {
	return {
		...collection,
		normalizedName: collection.name.normalize('NFKC').trim().toLocaleLowerCase()
	};
}
